File: src/lib/types/project.ts
export type ProjectType = "Research" | "Dev" | "Industry" | "Other";
export type ProjectStatus = "Pending Approval" | "Active" | "Completed" | "Rejected";
export type ProjectFlow = "internal" | "external";

export interface Faculty {
  id: string;
  name: string;
  email: string;
  department: string;
  institution?: string;
  role: "internal" | "external";
}

export interface AssignedStudent {
  id: string;
  name: string;
  rollNo: string;
  email: string;
  department: string;
}

export interface ExternalFacultyFormData {
  name: string;
  email: string;
  department: string;
  institution: string;
}

export interface Project {
  id: string;
  title: string;
  type: ProjectType;
  description: string;
  startDate: string;
  endDate: string;
  pptFile?: { name: string; url: string; uploadedAt: string; };
  status: ProjectStatus;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
  flow: ProjectFlow;                           // "internal" or "external"
  facultyCoordinator: Faculty;                 // 🏫 Internal coordinator (manages approvals & enrollment)
  externalFaculty?: Faculty;                   // 🌐 Outside faculty (optional, only for external flow)
  approvalSubmittedBy: string;                 // Always facultyCoordinator.id
  assignedStudents: AssignedStudent[];
  rejectionReason?: string;
  notes?: string;
}

export interface ProjectFormData {
  title: string;
  type: ProjectType;
  description: string;
  startDate: string;
  endDate: string;
  pptFile?: File;
  flow: ProjectFlow;                       // "internal" or "external"
  externalFaculty?: ExternalFacultyFormData; // Only for external flow
}

export interface ProjectFilters {
  status?: ProjectStatus;
  type?: ProjectType;
  searchTerm?: string;
  isExternalCollaboration?: boolean;
}

export interface ProjectsQueryParams extends ProjectFilters {
  page?: number;
  limit?: number;
}

File: backend/app/models/project.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import Column, DateTime, String, Text, JSON, ForeignKey
from app.db.base import Base


class Project(Base):
    __tablename__ = "projects"

    id = Column(String(50), primary_key=True, default=lambda: f"proj-{uuid.uuid4().hex[:8]}")
    title = Column(String(255), nullable=False)
    type = Column(String(50), nullable=False)  # Research, Dev, Industry, Other
    description = Column(Text, nullable=False)
    start_date = Column(String(20), nullable=False)
    end_date = Column(String(20), nullable=False)
    status = Column(String(50), nullable=False, default="Pending Approval")  # Pending Approval, Active, Completed, Rejected
    flow = Column(String(20), nullable=False, default="internal")  # internal or external
    ppt_file = Column(JSON, nullable=True)  # {name, url, uploadedAt}
    created_by = Column(String(150), nullable=False)
    created_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    faculty_coordinator = Column(JSON, nullable=False)  # {id, name, email, department, role} - internal coordinator
    external_faculty = Column(JSON, nullable=True)  # {id, name, email, department, role} - external faculty (only for external flow)
    approval_submitted_by = Column(String(150), nullable=False)  # always faculty_coordinator.id
    assigned_students = Column(JSON, nullable=True, default=[])  # [{id, name, rollNo, email, department}]
    rejection_reason = Column(Text, nullable=True)
    notes = Column(Text, nullable=True)
